pkg/client/userd/trafficmgr/traffic_manager.go

```diff
@@ -400,148 +410,130 @@ func (tm *TrafficManager) Run(c context.Context) error {
 	return tm.sessionInfo
 }
 
-// hasOwner parses an object and determines whether the object has an
-// owner that is of a kind we prefer. Currently the only owner that we
-// prefer is a Deployment, but this may grow in the future
-func (tm *TrafficManager) hasOwner(obj k8sapi.Object) bool {
-	for _, owner := range obj.GetOwnerReferences() {
-		if owner.Kind == "Deployment" {
-			return true
-		}
-	}
-	return false
-}
-
-// getReasonAndLabels gets the workload's associated labels, as well as a reason
-// it cannot be intercepted if that is the case.
-func (tm *TrafficManager) getReasonAndLabels(workload k8sapi.Workload) (labels map[string]string, reason string) {
-	labels = workload.GetPodTemplate().Labels
-	if workload.Replicas() == 0 {
-		reason = "Has 0 replicas"
-	}
-	return
-}
-
-// getInfosForWorkload creates a WorkloadInfo for every workload in names
-// of the given objectKind.  Additionally, it uses information about the
-// filter param, which is configurable, to decide which workloads to add
-// or ignore based on the filter criteria.
+// getInfosForWorkloads returns a list of workloads found in the given namespace that fulfils the given filter criteria.
 func (tm *TrafficManager) getInfosForWorkloads(
 	ctx context.Context,
-	workloads []k8sapi.Workload,
-	namespace string,
+	namespaces []string,
 	iMap map[string]*manager.InterceptInfo,
 	aMap map[string]*manager.AgentInfo,
 	filter rpc.ListRequest_Filter,
-) []*rpc.WorkloadInfo {
-	workloadInfos := make([]*rpc.WorkloadInfo, 0)
-	for _, workload := range workloads {
-		name := workload.GetName()
-		dlog.Debugf(ctx, "Getting info for %s %s.%s", workload.GetKind(), name, workload.GetNamespace())
-		iCept, ok := iMap[name]
-		if !ok && filter <= rpc.ListRequest_INTERCEPTS {
-			continue
+) ([]*rpc.WorkloadInfo, error) {
+	wis := make([]*rpc.WorkloadInfo, 0)
+	var err error
+	tm.wlWatcher.eachService(ctx, namespaces, func(svc *core.Service) {
+		var wls []k8sapi.Workload
+		if wls, err = tm.wlWatcher.findMatchingWorkloads(ctx, svc); err != nil {
+			return
 		}
-		agent, ok := aMap[name]
-		if !ok && filter <= rpc.ListRequest_INSTALLED_AGENTS {
-			continue
-		}
-		reason := ""
-		if agent == nil && iCept == nil {
-			labels, reason := tm.getReasonAndLabels(workload)
-			if reason == "" {
-				// If an object is owned by a higher level workload, then users should
-				// intercept that workload, so we will not include it in our slice.
-				if tm.hasOwner(workload) {
-					continue
-				}
-
-				matchingSvcs, err := install.FindMatchingServices(ctx, "", "", namespace, labels)
-				if err != nil {
-					continue
-				}
-				if len(matchingSvcs) == 0 {
-					reason = "No service with matching selector"
-				}
+		for _, workload := range wls {
+			name := workload.GetName()
+			dlog.Debugf(ctx, "Getting info for %s %s.%s, matching service %s.%s", workload.GetKind(), name, workload.GetNamespace(), svc.Name, svc.Namespace)
+			wlInfo := &rpc.WorkloadInfo{
```

Comment:
it'd be good if the service, with its UID, name and namespace were stored (could be optional) in the WorkloadInfo so that we can work our way back from service -> workload

Reply:
Agree. I'll add that in a future commit.

Reply:
it's ok i can take care of it, i think i'll need to make some changes to this code anyway @thallgren